File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/controller/CollectController.java
package com.windea.demo.cloudcollect.core.controller;

import com.windea.demo.cloudcollect.core.domain.entity.*;
import com.windea.demo.cloudcollect.core.domain.enums.CollectType;
import com.windea.demo.cloudcollect.core.domain.model.JwtUserDetails;
import com.windea.demo.cloudcollect.core.service.CollectService;
import io.swagger.annotations.*;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.security.Principal;
import java.util.Set;

/**
 * 收藏的控制器。
 */
@Api("收藏")
@RequestMapping("/collect")
@RestController
@CrossOrigin
public class CollectController {
	private final CollectService service;

	public CollectController(CollectService service) {
		this.service = service;
	}


	@ApiOperation("创建自己的收藏。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "collect", value = "新的收藏", required = true)
	})
	@PostMapping("/create")
	public Collect create(@RequestBody @Valid Collect collect, BindingResult bindingResult, Principal principal) {
		var user = ((JwtUserDetails) principal).getDelegateUser();
		return service.create(collect, user);
	}

	@ApiOperation("从别人的收藏创建自己的收藏。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "collect", value = "别人的收藏", required = true)
	})
	@PostMapping("/createFrom")
	public Collect createFrom(@RequestBody @Valid Collect collect, BindingResult bindingResult, Principal principal) {
		var user = ((JwtUserDetails) principal).getDelegateUser();
		return service.createFrom(collect, user);
	}

	@ApiOperation("删除自己的收藏。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path")
	})
	@DeleteMapping("/{id}")
	@PreAuthorize("hasPermission(#id,'Collect','delete')")
	public void delete(@PathVariable Long id) {
		service.delete(id);
	}

	@ApiOperation("修改自己的收藏。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path"),
		@ApiImplicitParam(name = "collect", value = "修改后的收藏", required = true)
	})
	@PutMapping("/{id}")
	@PreAuthorize("hasPermission(#id,'Collect','write')")
	public Collect modify(@PathVariable Long id, @RequestBody @Valid Collect collect, BindingResult bindingResult) {
		return service.modify(id, collect);
	}

	@ApiOperation("修改自己的收藏的分类。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path"),
		@ApiImplicitParam(name = "category", value = "修改后的收藏的分类", required = true)
	})
	@PutMapping("/{id}/category")
	@PreAuthorize("hasPermission(#id,'Collect','write')")
	public Collect modifyCategory(@PathVariable Long id, @RequestBody CollectCategory category) {
		return service.modifyCategory(id, category);
	}

	@ApiOperation("修改自己的收藏的标签。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path"),
		@ApiImplicitParam(name = "tags", value = "修改后的收藏的标签", required = true)
	})
	@PutMapping("/{id}/tags")
	@PreAuthorize("hasPermission(#id,'Collect','write')")
	public Collect modifyTags(@PathVariable Long id, @RequestBody Set<CollectTag> tags) {
		return service.modifyTags(id, tags);
	}

	@ApiOperation("修改自己的收藏的类型。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path"),
		@ApiImplicitParam(name = "type", value = "修改后的收藏的类型", required = true)
	})
	@PutMapping("/{id}/type")
	@PreAuthorize("hasPermission(#id,'CollectCategory','write')")
	public Collect modifyType(@PathVariable Long id, @RequestBody CollectType type) {
		return service.modifyType(id, type);
	}

	@ApiOperation("点赞某一收藏。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path")
	})
	@PutMapping("/{id}/praise")
	public Collect praise(@PathVariable Long id, Principal principal) {
		var user = ((JwtUserDetails) principal).getDelegateUser();
		return service.praise(id, user);
	}

	@ApiOperation("得到某一收藏。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path")
	})
	@GetMapping("/{id}")
	public Collect get(@PathVariable Long id) {
		return service.get(id);
	}

	@ApiOperation("分页得到某一收藏的所有点赞用户。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path"),
		@ApiImplicitParam(name = "pageable", value = "分页和排序", required = true)
	})
	@GetMapping("/{id}/praiseByUserPage")
	public Page<User> getPraiseByUserPage(@PathVariable Long id, @RequestParam Pageable pageable) {
		return service.getPraiseByUserPage(id, pageable);
	}

	@ApiOperation("得到某一收藏的点赞用户数量。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path")
	})
	@GetMapping("/{id}/praiseByUserCount")
	public Long getPraiseByUserCount(@PathVariable Long id) {
		return service.getPraiseByUserCount(id);
	}

	@ApiOperation("分页得到某一收藏的所有评论。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path"),
		@ApiImplicitParam(name = "pageable", value = "分页和排序", required = true)
	})
	@GetMapping("/{id}/commentPage")
	public Page<Comment> getCommentPage(@PathVariable Long id, @RequestParam Pageable pageable) {
		return service.getCommentPage(id, pageable);
	}

	@ApiOperation("得到某一收藏的评论数量。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path")
	})
	@GetMapping("/{id}/commentCount")
	public Long getCommentCount(@PathVariable Long id) {
		return service.getCommentCount(id);
	}

	@ApiOperation("分页得到所有收藏。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "pageable", value = "分页和排序", required = true)
	})
	@GetMapping("/findAll")
	@PreAuthorize("hasRole('ROLE_ADMIN')")
	public Page<Collect> findAll(@RequestParam Pageable pageable) {
		return service.findAll(pageable);
	}

	@ApiOperation("分页查询某一用户的所有已删除/未删除收藏。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "userId", value = "用户的id", required = true),
		@ApiImplicitParam(name = "deleteStatus", value = "删除状态", required = true),
		@ApiImplicitParam(name = "pageable", value = "分页和排序", required = true)
	})
	@GetMapping("/findByUserAndDeleteStatus")
	public Page<Collect> findByUserAndDeleteStatus(@RequestParam Long userId, @RequestParam Boolean deleteStatus,
		@RequestParam Pageable pageable) {
		return service.findByUserAndDeleteStatus(userId, deleteStatus, pageable);
	}

	@ApiOperation("根据名字分页模糊查询某一用户的所有收藏。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "userId", value = "用户的id", required = true),
		@ApiImplicitParam(name = "name", value = "名字", required = true),
		@ApiImplicitParam(name = "pageable", value = "分页和排序", required = true)
	})
	@GetMapping("/findByUserAndName")
	public Page<Collect> findByUserAndName(@RequestParam Long userId, @RequestParam String name,
		@RequestParam Pageable pageable) {
		return service.findByUserAndName(userId, name, pageable);
	}

	@ApiOperation("根据分类分页查询某一用户的所有收藏。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "categoryId", value = "分类的id", required = true),
		@ApiImplicitParam(name = "pageable", value = "分页和排序", required = true)
	})
	@GetMapping("/findByUserAndCategory")
	public Page<Collect> findByUserAndCategory(@RequestParam Long categoryId, @RequestParam Pageable pageable) {
		return service.findByUserAndCategory(categoryId, pageable);
	}

	@ApiOperation("根据标签分页查询某一用户的所有收藏。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "tagId", value = "标签的id", required = true),
		@ApiImplicitParam(name = "pageable", value = "分页和排序", required = true)
	})
	@GetMapping("/findByUserAndCategory")
	public Page<Collect> findByUserAndTag(@RequestParam Long tagId, @RequestParam Pageable pageable) {
		return service.findByUserAndTag(tagId, pageable);
	}

	@ApiOperation("根据类型分页查询某一用户的所有收藏。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "userId", value = "用户的id", required = true),
		@ApiImplicitParam(name = "type", value = "收藏的类型", required = true),
		@ApiImplicitParam(name = "pageable", value = "分页和排序", required = true)
	})
	@GetMapping("/findByUserAndType")
	public Page<Collect> findByUserAndType(@RequestParam Long userId, @RequestParam CollectType type,
		@RequestParam Pageable pageable) {
		return service.findByUserAndType(userId, type, pageable);
	}

	@ApiOperation("根据名字分页全局查询所有收藏。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "name", value = "名字", required = true),
		@ApiImplicitParam(name = "pageable", value = "分页和排序", required = true)
	})
	@GetMapping("/findByName")
	public Page<Collect> findByName(@RequestParam String name, @RequestParam Pageable pageable) {
		return service.findByName(name, pageable);
	}
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/controller/CollectController.kt
@file:Suppress("UNUSED_PARAMETER")

package com.windea.demo.cloudcollect.core.controller

import com.windea.demo.cloudcollect.core.domain.entity.*
import com.windea.demo.cloudcollect.core.enums.*
import com.windea.demo.cloudcollect.core.service.*
import com.windea.demo.cloudcollect.core.validation.group.*
import io.swagger.annotations.*
import org.springframework.data.domain.*
import org.springframework.security.access.prepost.*
import org.springframework.validation.*
import org.springframework.validation.annotation.*
import org.springframework.web.bind.annotation.*

@Api("收藏")
@RestController
@RequestMapping("/collect")
@CrossOrigin
class CollectController(
	private val collectService: CollectService
) {
	@ApiOperation("创建自己的收藏。")
	@PostMapping("/create")
	@PreAuthorize("isAuthenticated()")
	fun create(@RequestBody @Validated(Create::class) collect: Collect, bindingResult: BindingResult) {
		collectService.create(collect)
	}
	
	@ApiOperation("从别人的收藏创建自己的收藏。")
	@PostMapping("/fork")
	@PreAuthorize("isAuthenticated()")
	fun fork(@RequestBody collect: Collect) {
		collectService.fork(collect)
	}
	
	@ApiOperation("修改自己的收藏。")
	@PutMapping("/{id}")
	@PreAuthorize("isAuthenticated()")
	fun modify(@PathVariable id: Long, @RequestBody @Validated(Modify::class) collect: Collect, bindingResult: BindingResult) {
		collectService.modify(id, collect)
	}
	
	@ApiOperation("点赞某一收藏。")
	@PutMapping("/{id}/praise")
	@PreAuthorize("isAuthenticated()")
	fun praise(@PathVariable id: Long) {
		collectService.praise(id)
	}
	
	@ApiOperation("取消点赞某一收藏。")
	@PutMapping("/{id}/unpraise")
	@PreAuthorize("isAuthenticated()")
	fun unpraise(@PathVariable id: Long) {
		collectService.unpraise(id)
	}
	
	@ApiOperation("删除自己的收藏。")
	@DeleteMapping("/{id}")
	@PreAuthorize("isAuthenticated()")
	fun deleteById(@PathVariable id: Long) {
		collectService.deleteById(id)
	}
	
	@ApiOperation("根据id得到某一收藏。")
	@GetMapping("/{id}")
	fun findById(@PathVariable id: Long): Collect {
		return collectService.findById(id)
	}
	
	@ApiOperation("得到所有收藏。")
	@GetMapping("/findAll")
	fun findAll(pageable: Pageable): Page<Collect> {
		return collectService.findAll(pageable)
	}
	
	@ApiOperation("根据名字全局查询所有收藏。")
	@GetMapping("/findAllByNameContains")
	fun findAllByNameContains(@RequestParam name: String, pageable: Pageable): Page<Collect> {
		return collectService.findAllByNameContains(name, pageable)
	}
	
	@ApiOperation("根据分类id查询所有收藏。")
	@GetMapping("/findAllByCategoryId")
	fun findAllByCategoryId(@RequestParam categoryId: Long, pageable: Pageable): Page<Collect> {
		return collectService.findAllByCategoryId(categoryId, pageable)
	}
	
	@ApiOperation("根据分类名字全局查询所有收藏。")
	@GetMapping("/findAllByCategoryName")
	fun findAllByCategoryName(@RequestParam categoryName: String, pageable: Pageable): Page<Collect> {
		return collectService.findAllByCategoryName(categoryName, pageable)
	}
	
	@ApiOperation("根据分类名字全局模糊查询所有收藏。")
	@GetMapping("/findAllByCategoryNameContains")
	fun findAllByCategoryNameContains(@RequestParam categoryName: String, pageable: Pageable): Page<Collect> {
		return collectService.findAllByCategoryNameContains(categoryName, pageable)
	}
	
	@ApiOperation("根据标签id查询所有收藏。")
	@GetMapping("/findAllByTagId")
	fun findAllByTagId(@RequestParam tagId: Long, pageable: Pageable): Page<Collect> {
		return collectService.findAllByTagId(tagId, pageable)
	}
	
	@ApiOperation("根据标签名字全局查询所有收藏。")
	@GetMapping("/findAllByTagName")
	fun findAllByTagName(@RequestParam tagName: String, pageable: Pageable): Page<Collect> {
		return collectService.findAllByTagName(tagName, pageable)
	}
	
	@ApiOperation("根据标签名字全局模糊查询所有收藏。")
	@GetMapping("/findAllByTagNameContains")
	fun findAllByTagNameContains(@RequestParam tagName: String, pageable: Pageable): Page<Collect> {
		return collectService.findAllByTagNameContains(tagName, pageable)
	}
	
	@ApiOperation("根据用户id和收藏状态查询所有收藏。")
	@GetMapping("/findAllByUserId")
	fun findAllByUserId(@RequestParam userId: Long, pageable: Pageable): Page<Collect> {
		return collectService.findAllByUserId(userId, pageable)
	}
	
	@ApiOperation("根据名字和用户id模糊查询所有收藏。")
	@GetMapping("/findAllByNameContainsAndUserId")
	fun findAllByNameContainsAndUserId(@RequestParam name: String, @RequestParam userId: Long, pageable: Pageable): Page<Collect> {
		return collectService.findAllByNameContainsAndUserId(name, userId, pageable)
	}
	
	@ApiOperation("根据分类名字和用户id模糊查询所有收藏。")
	@GetMapping("/findAllByCategoryNameContainsAndUserId")
	fun findAllByCategoryNameContainsAndUserId(@RequestParam categoryName: String, @RequestParam userId: Long, pageable: Pageable): Page<Collect> {
		return collectService.findAllByCategoryNameContainsAndUserId(categoryName, userId, pageable)
	}
	
	@ApiOperation("根据标签名字和用户id模糊查询所有收藏。")
	@GetMapping("/findAllByTagNameContainsAndUserId")
	fun findAllByTagNameContainsAndUserId(@RequestParam tagName: String, @RequestParam userId: Long, pageable: Pageable): Page<Collect> {
		return collectService.findAllByTagNameContainsAndUserId(tagName, userId, pageable)
	}
	
	@ApiOperation("根据类型和用户id查询所有收藏。")
	@GetMapping("/findAllByTypeAndUserId")
	fun findAllByTypeAndUserId(@RequestParam type: CollectType, @RequestParam userId: Long, pageable: Pageable): Page<Collect> {
		return collectService.findAllByTypeAndUserId(type, userId, pageable)
	}
	
	@ApiOperation("根据点赞用户id查询所有收藏。")
	@GetMapping("/findAllByPraiseByUserId")
	fun findAllByPraiseByUserId(@RequestParam praiseByUserId: Long, pageable: Pageable): Page<Collect> {
		return collectService.findAllByPraiseByUserId(praiseByUserId, pageable)
	}
	
	@ApiOperation("根据名字和点赞用户id模糊查询所有收藏。")
	@GetMapping("/findAllByNameContainsAndPraiseByUserId")
	fun findAllByNameContainsAndPraiseByUserId(@RequestParam name: String, @RequestParam praiseByUserId: Long, pageable: Pageable): Page<Collect> {
		return collectService.findAllByNameContainsAndPraiseByUserId(name, praiseByUserId, pageable)
	}
	
	@ApiOperation("根据分类名字和点赞用户id模糊查询所有收藏。")
	@GetMapping("/findAllByCategoryNameContainsAndPraiseByUserId")
	fun findAllByCategoryNameContainsAndPraiseByUserId(@RequestParam categoryName: String, @RequestParam praiseByUserId: Long, pageable: Pageable): Page<Collect> {
		return collectService.findAllByCategoryNameContainsAndPraiseByUserId(categoryName, praiseByUserId, pageable)
	}
	
	@ApiOperation("根据类型和点赞用户id查询所有收藏。")
	@GetMapping("/findAllByTagNameContainsAndPraiseByUserId")
	fun findAllByTagNameContainsAndPraiseByUserId(@RequestParam tagName: String, @RequestParam praiseByUserId: Long, pageable: Pageable): Page<Collect> {
		return collectService.findAllByTagNameContainsAndPraiseByUserId(tagName, praiseByUserId, pageable)
	}
	
	@ApiOperation("根据标签名字和点赞用户id模糊查询所有收藏。")
	@GetMapping("/findAllByTypeAndPraiseByUserId")
	fun findAllByTypeAndPraiseByUserId(@RequestParam type: CollectType, @RequestParam praiseByUserId: Long, pageable: Pageable): Page<Collect> {
		return collectService.findAllByTypeAndPraiseByUserId(type, praiseByUserId, pageable)
	}
}
